astropy/stats/lombscargle/implementations/main.py

```diff
@@ -0,0 +1,322 @@
+"""
+Main Lomb-Scargle Implementation
+
+The ``lombscargle`` function here is essentially a sophisticated switch
+statement for the various implementations available in this submodule
+"""
+__all__ = ['lombscargle', 'available_methods']
+
+import warnings
+
+import numpy as np
+from astropy import units
+from astropy.utils.compat.numpy import broadcast_arrays
+
+from .slow_impl import lombscargle_slow
+from .fast_impl import lombscargle_fast
+from .scipy_impl import lombscargle_scipy
+from .chi2_impl import lombscargle_chi2
+from .fastchi2_impl import lombscargle_fastchi2
+from .cython_impl import lombscargle_cython
+
+
+METHODS = {'slow': lombscargle_slow,
+           'fast': lombscargle_fast,
+           'chi2': lombscargle_chi2,
+           'scipy': lombscargle_scipy,
+           'fastchi2': lombscargle_fastchi2,
+           'cython': lombscargle_cython}
+
+
+def available_methods():
+    methods = ['auto', 'slow', 'chi2', 'cython']
+
+    # Numpy 1.8 or newer required for fast algorithms
+    if hasattr(np.ufunc, 'at'):
+        methods.extend(['fast', 'fastchi2'])
+
+    try:
+        import scipy
+        methods.append('scipy')
+    except ImportError:
+        pass
+    return methods
+
+
+def validate_inputs(t, y, dy=None, frequency=None, strip_units=True):
+    """Validation of input shapes & units
+
+    This utility function serves a few purposes:
+
+    - it validates that the shapes of t, y, and dy match, and broadcasts
+      them to a common 1D shape
+    - if any of t, y, day, or frequency are astropy Quantities (i.e. have
+      units attached), it validates that the units are compatible, and does
+      any necessary unit conversions
+    - if ``strip_units == True``, it strips units from all the arrays
+      before returning them.
+    - all relevant units are returned in ``unit_dict``
+
+    Parameters
+    ----------
+    t, y : array_like or Quantity
+    dy, frequency : array_like or Quantity (optional)
+    strip_units : bool (optional, default=True)
+        if True, the returned quantities will have units stripped.
+
+    Returns
+    -------
+    t, y, dy, frequency : ndarray, Quantity, or None
+        reshaped and/or unit-stripped arrays
+    unit_dict : dict
+        dictionary of relevant units
+    """
+    if dy is None:
+        t, y = broadcast_arrays(t, y, subok=True)
+    else:
+        t, y, dy = broadcast_arrays(t, y, dy, subok=True)
+
+    if t.ndim != 1:
+        raise ValueError("Input times & data must be one-dimensional")
+
+    has_units = any(isinstance(arr, units.Quantity)
+                    for arr in (t, y, dy, frequency))
+
+    if has_units:
+        power_unit = units.dimensionless_unscaled
+
+        t = units.Quantity(t)
```

Comment:
Add `copy=False` to these initialisations, so that if something is already OK, it doesn't get unnecessarily copied (this makes it similar to `np.asarray`).  Of course, the moment conversions are needed, copies will get made.